src/nifs/protogalaxy/poly.rs

```diff
@@ -0,0 +1,178 @@
+use std::num::NonZeroU32;
+
+use ff::Field;
+use halo2curves::CurveAffine;
+use itertools::Itertools;
+use rayon::prelude::*;
+use tracing::*;
+
+use crate::{
+    fft,
+    plonk::{self, eval, GetChallenges, GetWitness, PlonkStructure},
+    polynomial::{lagrange, univariate::UnivariatePoly},
+};
+
+use super::pow_i;
+
+#[derive(Debug, thiserror::Error, PartialEq, Eq, Clone)]
+pub enum Error {
+    #[error(transparent)]
+    Eval(#[from] eval::Error),
+    #[error("pow_i iterator ended before witnesses")]
+    PowiEndedBeforeWitnesses,
+}
+
+/// Analog of [`itertools::structs::MultiProduct`], but:
+/// - without [`Clone`] require
+/// - without logic for diff len of iterators
+struct MultiProduct<Item, Iter>
+where
+    Iter: Iterator<Item = Item>,
+{
+    iterators: Box<[Iter]>,
+}
+
+impl<Item, Iter> Iterator for MultiProduct<Item, Iter>
+where
+    Iter: Iterator<Item = Item>,
+{
+    type Item = Box<[Item]>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.iterators.iter_mut().map(|p| p.next()).collect()
+    }
+}
+
+#[instrument(skip_all)]
+pub(crate) fn compute_F<C: CurveAffine>(
+    beta: C::ScalarExt,
+    delta: C::ScalarExt,
+    S: &PlonkStructure<C::ScalarExt>,
+    trace: &(impl Sync + GetChallenges<C::ScalarExt> + GetWitness<C::ScalarExt>),
+) -> Result<UnivariatePoly<C::ScalarExt>, Error> {
+    let count_of_rows = 2usize.pow(S.k as u32);
+    let count_of_gates = S.gates.len();
+
+    let count_of_evaluation = count_of_rows * count_of_gates;
+    let log_n = NonZeroU32::new(count_of_evaluation.ilog2()).unwrap();
+
+    let points_count = count_of_evaluation
+        .next_power_of_two()
+        .ilog2()
+        .next_power_of_two() as usize;
+
+    debug!(
+        "
+        count_of_rows: {count_of_rows};
+        count_of_gates: {count_of_gates};
+        count_of_evaluation: {count_of_evaluation};
+        log_n: {log_n};
+        points_count: {points_count}"
+    );
+
+    let mut evaluated_points = MultiProduct {
+        iterators: lagrange::iter_cyclic_subgroup::<C::ScalarExt>(log_n.get())
```

Comment:
```
let n = points_count;
let log_n = n.ilog2();
```
so we have the size of cyclic group is exactly n. If we pick a larger than n cyclic group here and only take first n elements, the ifft will not give correct result.

Reply:
add a TODO for O(N) size algorithm here

Reply:
What TODO? We calculate one point by O(n), but we have log(n) points

Reply:
`log_n` is certainly a bad name (but better `t`) and we definitely need a section of notation for each thing within the protocol. Using letters in the code is always a compromise, but here I can feel how in six months we won't be able to figure out what `n` is.

Reply:
So for O(N), I should say exactly N operation without extra overhead. See the comment below.